File: addon/Familiar.lua
local _, WS = ...

-- Familiars: spirit wolves that HUNT. Each wolf picks the nearest enemy,
-- bounds it down, and bites - a leaping melee maul - then peels off, retargets,
-- and does it again. Rendered with the real spirit-wolf 3D model (icon
-- fallback), bounding as it runs and pouncing on each bite.
--
-- Granted in-run by the Spirit Companion passive (one wolf per rank). Numbers
-- live in `Familiar.tuning` so the Tuning page can adjust them live.
WS.Familiar = {}
local Familiar = WS.Familiar

local MAX = 6 -- shared cap across every summon kind (3 wolves + 3 ghouls)
local SIZE = 46

Familiar.tuning = {
    dashSpeed   = 340,  -- how fast a wolf closes on prey
    biteRadius  = 46,   -- area of each bite
    biteCooldown = 0.55,-- seconds between bites
    dmgBase     = 14,   -- bite damage floor
    dmgPerLevel = 1.0,  -- + per survivor level
    huntRange   = 560,  -- how far a wolf will chase
    leash       = 440,  -- snap back toward the survivor past this
    wanderSpeed = 150,  -- loping speed when nothing to hunt
    -- Ghouls are the DK's shambling counterpart: slower to close and slower to
    -- swing, but they hit harder and their claws sweep a wider arc.
    ghoulSpeedMult  = 0.70, -- x dashSpeed
    ghoulRadiusMult = 1.35, -- x biteRadius
    ghoulCdMult     = 1.45, -- x biteCooldown
    ghoulDmgMult    = 2.10, -- x bite damage
    -- Creature models. Blind display-ID picks are unreliable, so these are
    -- tunable and previewable live with "/survivors summon <kind> <id>".
    -- 1236 = spirit wolf. 137 = the classic ghoul, the same model the
    -- "Rotting Shambler" enemy already uses, so it is known-good here.
    wolfDisplay  = 1236,
    ghoulDisplay = 137,
}

-- Per-kind presentation + the multipliers applied on top of the shared tuning.
-- Adding another summon later means one more entry here, nothing else.
local KINDS = {
    wolf = {
        display = "wolfDisplay", icon = "Interface\\Icons\\Spell_Nature_SpiritWolf",
        source = "wolves", flash = { 0.60, 0.85, 1.00 },
        -- (wolves use the shared tuning values unmodified)
    },
    ghoul = {
        display = "ghoulDisplay", icon = "Interface\\Icons\\Spell_Shadow_AnimateDead",
        source = "ghouls", flash = { 0.55, 0.90, 0.40 },
        speedMult = "ghoulSpeedMult", radiusMult = "ghoulRadiusMult",
        cdMult = "ghoulCdMult", dmgMult = "ghoulDmgMult",
    },
}
Familiar.KINDS = KINDS

function Familiar:Initialize()
    self.list = {}
    self.count = 0
end

function Familiar:Reset()
    self.count = 0
    if self.list then
        for i = 1, #self.list do
            local fam = self.list[i]
            fam.frame:Hide()
            if fam.model then fam.model:Hide() end
        end
    end
end

-- Adds a summon of `kind` ("wolf" / "ghoul"), rendered as a 3D creature model
-- (icon fallback if it fails).
function Familiar:Add(kind)
    local spec = KINDS[kind] or KINDS.wolf
    local displayId = self.tuning[spec.display]
    if (self.count or 0) >= MAX then return end
    self.count = (self.count or 0) + 1
    local fam = self.list[self.count]
    if not fam then
        local frame = WS.UI:CreateEntityFrame(SIZE)
        WS.UI:ApplyRoundMask(frame)
        frame:SetFrameLevel(WS.UI.world:GetFrameLevel() + 12)
        fam = { frame = frame }
        -- The 3D wolf sits above the (fallback) icon.
        local model = CreateFrame("PlayerModel", nil, frame)
        model:SetPoint("TOPLEFT", -8, 16)
        model:SetPoint("BOTTOMRIGHT", 8, -4)
        model:SetScript("OnModelLoaded", function()
            if fam.modelActive then frame.icon:SetAlpha(0) end
        end)
        fam.model = model
        self.list[self.count] = fam
    end

    fam.spec = spec
    fam.frame.icon:SetTexture(spec.icon)
    fam.frame.icon:SetTexCoord(0.08, 0.92, 0.08, 0.92)
    fam.frame.icon:SetAlpha(1)
    fam.frame:SetSize(SIZE, SIZE)

    fam.modelActive = false
    fam.model:ClearModel()
    if displayId then
        fam.model:SetDisplayInfo(displayId)
        fam.model:SetAlpha(1)
        fam.model:Show()
        fam.modelActive = true
        fam.facing = 0
        fam.model:SetFacing(0)
    end

    local player = WS.Game.player
    local a = (self.count / MAX) * WS.tau
    fam.x = player.x + WS.cos(a) * 60
    fam.y = player.y + WS.sin(a) * 60
    fam.target = nil
    fam.biteTimer = 0
    fam.leap = 0        -- >0 during a pounce (drives the bound arc + scale pop)
    fam.wander = a
    fam.frame:Show()
end

function Familiar:Update(dt)
    if not self.count or self.count == 0 then return end
    local player = WS.Game.player
    local t = self.tuning

    -- Unholy Command (and any future summon buff) scales every kind at once.
    local dmgBoost = 1 + (player.summonDamage or 0)
    local hasteBoost = 1 + (player.summonHaste or 0)

    for i = 1, self.count do
        local fam = self.list[i]
        local spec = fam.spec or KINDS.wolf
        -- Per-kind multipliers over the shared tuning (wolves = 1x across the board).
        local kSpeed = spec.speedMult and t[spec.speedMult] or 1
        local kRadius = spec.radiusMult and t[spec.radiusMult] or 1
        local kCd = spec.cdMult and t[spec.cdMult] or 1
        local kDmg = spec.dmgMult and t[spec.dmgMult] or 1
        fam.biteTimer = fam.biteTimer - dt
        if fam.leap > 0 then fam.leap = fam.leap - dt end

        -- Drop a dead / gone / too-far target.
        local target = fam.target
        if target and (not target.poolIndex
            or WS.DistanceSquared(fam.x, fam.y, target.x, target.y) > t.huntRange * t.huntRange) then
            target = nil
        end
        if not target and fam.biteTimer <= 0 then
            target = WS.Enemy:FindNearest(fam.x, fam.y, t.huntRange)
        end
        fam.target = target

        local speed, tx, ty, moving = 0, nil, nil, true
        if target then
            tx, ty, speed = target.x, target.y, t.dashSpeed * kSpeed
        else
            fam.wander = fam.wander + dt * 1.6
            tx = player.x + WS.cos(fam.wander) * 90
            ty = player.y + WS.sin(fam.wander) * 90
            speed = t.wanderSpeed
        end
        if WS.DistanceSquared(fam.x, fam.y, player.x, player.y) > t.leash * t.leash then
            tx, ty, speed, fam.target = player.x, player.y, t.dashSpeed * kSpeed, nil
        end

        local dx, dy, dist = WS.Normalize(tx - fam.x, ty - fam.y)
        fam.x = fam.x + dx * speed * dt
        fam.y = fam.y + dy * speed * dt
        if speed < 20 then moving = false end

        -- Drive the model: face travel, run while moving, and leap on bite.
        if fam.modelActive then
            local facing = (dx < -0.05 and 0.7) or (dx > 0.05 and -0.7) or fam.facing or 0
            if facing ~= fam.facing then fam.facing = facing; fam.model:SetFacing(facing) end
            local anim = (fam.leap > 0 and 5) or (moving and 5 or 0)
            if anim ~= fam.anim then fam.anim = anim; fam.model:SetAnimation(anim) end
        end

        -- Bounding: a bob while running, and a bigger arc mid-pounce.
        local bob = 0
        if fam.leap > 0 then
            bob = WS.sin((1 - fam.leap / 0.28) * WS.pi) * 22 -- pounce arc
        elseif moving then
            bob = WS.abs(WS.sin(WS.Game.run.time * 13 + i)) * 6 -- running bound
        end
        local size = SIZE + (fam.leap > 0 and 10 or 0)
        WS.UI:Place(fam.frame, fam.x, fam.y + bob, size)

        -- Strike when it reaches prey: a wolf's leaping maul, a ghoul's wide claw.
        local radius = t.biteRadius * kRadius
        if target and fam.biteTimer <= 0 and dist < target.radius + radius * 0.6 then
            fam.biteTimer = t.biteCooldown * kCd / hasteBoost
            fam.leap = 0.28
            local dmg = (t.dmgBase + player.level * t.dmgPerLevel) * kDmg * dmgBoost
                * player.damageMultiplier * WS.Constants.PLAYER_DAMAGE_SCALE
            WS.Enemy:DamageArea(fam.x, fam.y, radius, dmg, nil, nil, spec.source)
            WS.Projectile:SpawnFlash(fam.x, fam.y, radius, spec.flash)
            fam.target = nil
        end
    end
end

-- Re-skin every summon of `kind` on the spot. Display IDs are near-impossible to
-- pick blind, so "/survivors summon ghoul 137" swaps the model live rather than
-- forcing a new run to find out what a number looks like.
function Familiar:SetKindDisplay(kind, displayId)
    local spec = KINDS[kind]
    if not spec then return false end
    self.tuning[spec.display] = displayId
    for i = 1, (self.count or 0) do
        local fam = self.list[i]
        if fam and fam.spec == spec and fam.model then
            fam.model:ClearModel()
            fam.model:SetDisplayInfo(displayId)
            fam.model:SetAlpha(1)
            fam.model:Show()
            fam.modelActive = true
            fam.facing, fam.anim = nil, nil -- force facing/animation to re-apply
        end
    end
    return true
end

function Familiar:Clear()
    self:Reset()
end

File: src/game/familiar.js
/* Familiars: summons that HUNT. Each picks the nearest enemy, bounds it down,
 * bites, then peels off and retargets. Spirit wolves are fast and light;
 * ghouls are slower to close and slower to swing, but hit far harder with a
 * wider sweep. Granted in-run by Spirit Companion / Grave Call. */
'use strict';
(function (WS) {

  const MAX = 6;   // shared cap across every summon kind
  /* A leashed summon is only released once it is back inside this fraction of
     the leash, and it only takes marks inside this fraction of it. Both exist
     to keep the come-home and the go-hunt decisions from meeting at a single
     pixel and arguing there forever. */
  const RECALL = 0.55;
  const HUNT_IN = 0.90;

  const Familiar = {
    list: [],
    tuning: {
      dashSpeed: 340,
      biteRadius: 46,
      biteCooldown: 0.55,
      dmgBase: 14,
      dmgPerLevel: 1.0,
      huntRange: 560,
      leash: 440,
      wanderSpeed: 150,
      ghoulSpeedMult: 0.70,
      ghoulRadiusMult: 1.35,
      ghoulCdMult: 1.45,
      ghoulDmgMult: 2.10,
    },
    KINDS: {
      wolf: { art: 'spiritwolf', tint: [0.60, 0.85, 1.00], source: 'wolves' },
      ghoul: {
        art: 'risen', tint: [0.55, 0.90, 0.40], source: 'ghouls',
        speedMult: 'ghoulSpeedMult', radiusMult: 'ghoulRadiusMult',
        cdMult: 'ghoulCdMult', dmgMult: 'ghoulDmgMult',
      },
    },
  };

  Familiar.init = function () { this.list = []; };
  Familiar.reset = function () { this.list.length = 0; };

  Familiar.add = function (kind) {
    if (this.list.length >= MAX) return;
    const player = WS.Game.player;
    const spec = this.KINDS[kind] || this.KINDS.wolf;
    this.list.push({
      kind, spec,
      x: player.x + WS.randRange(-50, 50),
      y: player.y + WS.randRange(-50, 50),
      target: null,
      biteTimer: 0,
      pounce: 0,
      leashed: false,
      bob: WS.random() * WS.TAU,
      facing: 1,
    });
    WS.FX.flash(player.x, player.y, 60, spec.tint, 0.4);
  };

  Familiar.update = function (dt) {
    const player = WS.Game.player;
    const t = this.tuning;

    for (const fam of this.list) {
      const spec = fam.spec;
      const speedMult = spec.speedMult ? t[spec.speedMult] : 1;
      const radiusMult = spec.radiusMult ? t[spec.radiusMult] : 1;
      const cdMult = spec.cdMult ? t[spec.cdMult] : 1;
      const dmgMult = spec.dmgMult ? t[spec.dmgMult] : 1;

      /* Leash, with hysteresis: once called off, a summon comes properly home
         before it hunts again.
         
         Without the second threshold this was a one-frame limit cycle and it
         pinned summons in place. Crossing the leash cleared the target and
         sent the summon home; one frame later it was a pixel inside the leash,
         reacquired the same mark, dashed out, and was called off again.
         Traced, a wolf sat between 434 and 445 pixels from the survivor - the
         leash is 440 - with NO target for 1.6 seconds while a creature stood
         148 pixels away.
         
         It shows up worst during a timestop, which is what was reported: the
         leash is anchored to the survivor, so walking drags the summon out of
         the cycle on its own. Stand still - which is exactly what a timestop
         invites you to do - and nothing breaks it. */
      const homeDist = WS.dist(fam.x, fam.y, player.x, player.y);
      if (homeDist > t.leash) fam.leashed = true;
      else if (homeDist < t.leash * RECALL) fam.leashed = false;

      /* Hunt around the SURVIVOR, not around the summon. Measured from the
         summon, a wolf already out at the leash could commit to a mark another
         huntRange beyond it - a mark it is structurally forbidden to reach.
         Bounded by the leash, everything it can see is something it can get
         to. */
      const reachable = WS.min(t.huntRange, t.leash * HUNT_IN);
      if (fam.leashed) {
        fam.target = null;
      } else if (!fam.target || fam.target._dead
        || WS.dist(player.x, player.y, fam.target.x, fam.target.y) > reachable) {
        fam.target = WS.Enemy.findNearest(player.x, player.y, reachable);
      }

      let tx, ty, speed;
      if (fam.leashed) {
        [tx, ty] = WS.normalize(player.x - fam.x, player.y - fam.y);
        speed = t.dashSpeed * speedMult;
      } else if (fam.target) {
        [tx, ty] = WS.normalize(fam.target.x - fam.x, fam.target.y - fam.y);
        speed = t.dashSpeed * speedMult * (1 + player.summonHaste * 0.25);
      } else {
        // Nothing to hunt: lope around the survivor.
        const a = fam.bob;
        const ox = player.x + WS.cos(a) * 70, oy = player.y + WS.sin(a) * 70;
        [tx, ty] = WS.normalize(ox - fam.x, oy - fam.y);
        speed = t.wanderSpeed;
      }
      fam.x += tx * speed * dt;
      fam.y += ty * speed * dt;
      if (tx !== 0) fam.facing = tx < 0 ? -1 : 1;
      fam.bob += dt * 3;
      if (fam.pounce > 0) fam.pounce -= dt;

      fam.biteTimer -= dt * (1 + player.summonHaste);
      if (fam.biteTimer <= 0 && fam.target) {
        const reach = t.biteRadius * radiusMult * 0.6 + fam.target.radius;
        if (WS.dist2(fam.x, fam.y, fam.target.x, fam.target.y) <= reach * reach) {
          fam.biteTimer = t.biteCooldown * cdMult;
          fam.pounce = 0.18;
          const damage = (t.dmgBase + t.dmgPerLevel * player.level)
            * dmgMult
            * player.damageMultiplier
            * (1 + player.summonDamage)
            * WS.CONST.PLAYER_DAMAGE_SCALE;
          WS.Enemy.damageArea(fam.x, fam.y, t.biteRadius * radiusMult * player.areaMultiplier,
            damage, null, null, spec.source);
          WS.FX.flash(fam.x, fam.y, t.biteRadius * radiusMult * 0.8, spec.tint, 0.2);
        }
      }
    }
  };

  WS.Familiar = Familiar;

})(window.WS);
